Let _admin role bypass VDU

nickva opened the pull request:
This is necessary in the case where an upgrade happens and replicator had
invalid document in them. Replicator would read it and then try to update it
with a failed state. However update will fail because the VDU might prevent it.

rnewson commented:
Can we tighten this? We don't want admins to be generally free to write garbage docs. I suggest we allow invalid docs (missing "source", for example) only when there's a _replication_state of "failed".

nickva commented:
@rnewson That's a good idea. Agreed

nickva committed:
Let _admin role bypass VDU for failed docs

This is necessary in the case where an upgrade happens and replicator had
invalid document in them. Replicator would read it and then try to update it
with a failed state. However update will fail because the VDU might prevent it.

nickva committed:
Provide better feedback for a bad replicator database.

Previously inserting a bad replicator doc with a missing source or target would
result in a function_clause error without an explantion what the function was.

Turn that into an error message that's more useful.

Example:

```
 {u'_id': u'bad',
  u'_replication_state': u'failed',
  u'_replication_state_reason': u'Missing replicator database',
  u'_replication_state_time': u'2016-10-14T14:22:37-04:00',
  u'_rev': u'8-c11b9ce3eb5730baa60e8335e0b22035',
  u'blah': u'x',
  u'owner': u'adm'}
```

rnewson commented on src/couch_replicator_js_functions.hrl:
```diff
@@ -53,6 +53,10 @@
         var isReplicator = (userCtx.roles.indexOf('_replicator') >= 0);
         var isAdmin = (userCtx.roles.indexOf('_admin') >= 0);
 
+        if (isAdmin && newDoc._replication_state === 'failed') {
```
no need to test for admin

rnewson commented on src/couch_replicator_js_functions.hrl:
```diff
@@ -53,6 +53,10 @@
         var isAdmin = (userCtx.roles.indexOf('_admin') >= 0);
 
+        if (isAdmin && newDoc._replication_state === 'failed') {
+            return;
```
add a comment saying we're explicitly skipping validation.

rnewson commented on src/couch_replicator_js_functions.hrl:
```diff
@@ -53,6 +53,14 @@
+        if (newDoc._replication_state === 'failed') {
+            // Skip validation in case when we update the document with the
+            // failed state. In this case it might be malformed. However,
+            // replicator will not pay attention or failed documents so this
```
"attention _to_ failed"

rnewson commented:
+1
